launch_testing/examples/test_pub_sub_launch.py

```diff
@@ -0,0 +1,230 @@
+    # once (40s total timeout).
+    lifecycle_pub_sub_launch_description.add_action(
+        AssertSequenceOnce([
+            variables.EventEmitted(
+                event_type=lifecycle.ChangeState
+            ),
+            variables.EventEmitted(
+                event_type=ProcessStarted
+            )
+        ], timeout=[30, 10])
+    )
+
+    ros_test_launch_service.include_launch_description(
+        lifecycle_pub_sub_launch_description
+    )
+    # TODO(hidmic): implement launch_testing specific pytest plugin to
+    # aggregate all test result information at the launch system-level
+    # and below.
+    assert ros_test_launch_service.run() == 0
+
+
+def test_pub_sub(ros_test_launch_service):
+    """
+    Tests a basic pub/sub setup for proper output matching.
+    """
+    ld = launch.LaunchDescription()
+    talker_node_action = Execute(
+        launch_ros.actions.Node(
+            package='demo_nodes_cpp',
+            node_executable='talker',
+            remappings=[('chatter', 'my_chatter')]
+        ), output='screen'
+    )
+    ld.add_action(talker_node_action)
+    listener_node_action = Execute(
+        launch_ros.actions.Node(
+            package='demo_nodes_py',
+            node_executable='listener',
+            remappings=[('chatter', 'my_chatter')]
+        ),  output='screen'
+    )
+    ld.add_action(listener_node_action)
+    # Asserts that not a single talker screen output
+    # line contains the ERROR string throughout the duration
+    # of the test (i.e. it completes when all other tests
+    # have completed). The assertion is strict (i.e. failure
+    # on a single output event means failure altogether).
+    ld.add_action(
+        Assert(not predicates.regex_match(
+            variables.Output(talker_node_action),
+            pattern='.*ERROR.*', flags=None
+        ), timeout=None, strict=True)
+    )
+    # Asserts that a the listener outputs 'I heard...'
+    # at least once in the first 10s of the test. The
+    # assertion is not strict (i.e. failure on a single
+    # output event does not mean failure altogether).
+    ld.add_action(
+        Assert(predicates.count(predicates.regex_match(
+            variables.Output(listener_node_action),
+            pattern='I heard.*', flags=None
+        )) > 0, message='No message received!', timeout=10)
+    )
+    ros_test_launch_service.include_launch_description(ld)
+
+    # TODO(hidmic): implement launch_testing specific pytest plugin to
+    # aggregate all test result information at the launch system-level
+    # and below.
+    assert ros_test_launch_service.run() == 0
+
+
+def test_intra_comm_pub_sub(ros_test_launch_service):
+    """
+    Tests a intra process communication between composable
+    pub/sub nodes via gtest.
+    """
+    # Launches a gtest that happens to be a container
+    # for composable nodes.
+    ld = LaunchDescription([
+        GTest(ComposableNodeProcess(
```

Comment:
I don't see how a composable node process can be a gtest. Which part is actually a gtest in this configuration?

Reply:
Here's the use case I came up with. 

Say you want to test a pointcloud postprocessing pipeline that runs in a single process through composable nodes for improved performance. Testing from outside the process may yield different results, specially if the pipeline is time critical. So the test has to be run from within the process.

In that scenario, I envisioned a container process that's actually a gtest with a fixture that takes care of pipeline setup and teardown. With that in mind composition (actually, it's closer to decoration) instead of duplication makes sense because the two traits of the executable (being a gtest and being composable node) are orthogonal.